### PR TITLE
Move Loader metadata parsing out of ctor

### PR DESCRIPTION
**Pull request by Kh4L**
- Factorizing Loader init out of constructor
- Adding Reader/Loader option to make the init lazy: executed only on the first Run
- Moves COCOReader data preparation logic in COCOLoader

Co-author @awolant 

Moved to #746

### CODE REVIEW

File: dali/pipeline/operators/reader/loader/loader.h [hunk 1]
```diff
@@ -59,14 +59,18 @@ class Loader {
...
+      lazy_(options.GetArgument<bool>("lazy")) {
     DALI_ENFORCE(initial_empty_size_ > 0, "Batch size needs to be greater than 0");
     DALI_ENFORCE(num_shards_ > shard_id_, "num_shards needs to be greater than shard_id");
     // initialize a random distribution -- this will be
     // used to pick from our sample buffer
     dis = std::uniform_int_distribution<>(0, initial_buffer_fill_);
     std::seed_seq seq({seed_});
     e_ = std::default_random_engine(seq);
+    if (!lazy_) {
+      PrepareMetadata();
```
Comment (Kh4L) [inline dali/pipeline/operators/reader/loader/loader.h]: Actually doesn't make sense to call a virtual method in the ctor, to rework

File: dali/pipeline/operators/reader/loader/file_loader.h [hunk 2]
```diff
@@ -51,11 +51,17 @@ class FileLoader : public Loader<CPUBackend, ImageLabelWrapper> {
...
       current_index_(0),
       current_epoch_(0) {
-    file_list_ = spec.GetArgument<string>("file_list");
+  }
+
+  void PrepareEmpty(ImageLabelWrapper &tensor) override;
+  void ReadSample(ImageLabelWrapper &tensor) override;
+
+  void PrepareMetadata() override {
     mmap_reserver = FileStream::FileStreamMappinReserver(
```
Comment (awolant) [inline dali/pipeline/operators/reader/loader/file_loader.h]: Maybe move this into constructor?

File: dali/pipeline/operators/reader/loader/loader.h [hunk 3]
```diff
@@ -71,6 +72,14 @@ class Loader {
 
   virtual ~Loader() = default;
 
+  // We need this two stage init because overriden PrepareMetadata
+  // is not known in Loader ctor
+  void Init() {
+    if (!lazy_) {
+      PrepareMetadata();
```
Comment (awolant) [inline dali/pipeline/operators/reader/loader/loader.h]: Maybe wrap this with `std::cal_once` also?

File: dali/pipeline/operators/reader/reader_op.h [hunk 4]
```diff
@@ -168,6 +168,13 @@ class DataReader : public Operator<Backend> {
     return sample_ptr;
   }
 
+  template<typename T, typename... Args>
+  std::unique_ptr<T> InitLoader(const OpSpec& spec, Args... args) {
```
Comment (awolant) [inline dali/pipeline/operators/reader/reader_op.h]: Maybe this could be free function?

File: dali/pipeline/operators/reader/loader/loader.h (hunk 1 above)
Comment (Kh4L) [reply dali/pipeline/operators/reader/loader/loader.h]: Done

File: dali/pipeline/operators/reader/loader/file_loader.h (hunk 2 above)
Comment (Kh4L) [reply dali/pipeline/operators/reader/loader/file_loader.h]: Done

File: dali/pipeline/operators/reader/loader/loader.h (hunk 3 above)
Comment (Kh4L) [reply dali/pipeline/operators/reader/loader/loader.h]: Agreed ;)

File: dali/pipeline/operators/reader/reader_op.h (hunk 4 above)
Comment (Kh4L) [reply dali/pipeline/operators/reader/reader_op.h]: Agreed ;)

Comment (Kh4L) [reply dali/pipeline/operators/reader/reader_op.h]: Done

**Comment by Kh4L** [comment]
Build 695013

File: dali/pipeline/operators/reader/loader/loader.h [hunk 5]
```diff
@@ -59,7 +59,8 @@ class Loader {
       read_ahead_(options.GetArgument<bool>("read_ahead")),
       stick_to_shard_(options.GetArgument<bool>("stick_to_shard")),
       device_id_(options.GetArgument<int>("device_id")),
-      skip_cached_images_(options.GetArgument<bool>("skip_cached_images")) {
+      skip_cached_images_(options.GetArgument<bool>("skip_cached_images")),
+      lazy_(options.GetArgument<bool>("lazy")) {
```
Comment (jantonguirao) [inline dali/pipeline/operators/reader/loader/loader.h]: Knitpick: maybe something more descriptive (`lazy_init` or something)

File: dali/pipeline/operators/reader/loader/loader.h [hunk 6]
```diff
@@ -229,11 +247,23 @@ class Loader {
...
+  bool lazy_;
+
   // Image cache
   std::once_flag fetch_cache_;
   std::shared_ptr<ImageCache> cache_;
 };
 
+template<typename T, typename... Args>
+std::unique_ptr<T> InitLoader(const OpSpec& spec, Args... args) {
+  std::unique_ptr<T> l (new T(spec, args...));
```
Comment (jantonguirao) [inline dali/pipeline/operators/reader/loader/loader.h]: knitpick
```suggestion
  std::unique_ptr<T> loader (new T(spec, args...));
```

**Comment by jantonguirao** [comment]
@Kh4L 
We are missing a `DALI_ENFORCE` when calling reader size with lazy initialization enabled

File: dali/pipeline/operators/reader/loader/loader.h (hunk 5 above)
Comment (Kh4L) [reply dali/pipeline/operators/reader/loader/loader.h]: Resolve in #746

File: dali/pipeline/operators/reader/loader/loader.h (hunk 6 above)
Comment (Kh4L) [reply dali/pipeline/operators/reader/loader/loader.h]: Resolve in #746

**Comment by Kh4L** [comment]
Closing in favor of #746 that contains the changes and more.

**Pull request closed without merging**